rpcs3/Emu/RSX/RSXFIFO.cpp

```diff
@@ -144,32 +147,41 @@ namespace rsx
 								break;
 							}
 
+							bytes_read += 128;
 							continue;
 						}
 					}
 
 					if (!start_time)
 					{
+						if ((bytes_read << 1) >= m_cache_size)
```

Comment:
I suggest avoiding making accurate code inaccurate, there was another pr which eliminated this problem altogether..

Reply:
Games don'e use PUTLLC to update FIFO which is why I made the pr https://github.com/RPCS3/rpcs3/pull/12183.

Reply:
How is it inaccurate?

Reply:
RSX prefetches its data for up to 1024 bytes, I have already seen a game using this feature otherwise it crashes with FIFO errors.

Reply:
The performance issue comes from that reservations locks are shared (there are only 512 allocated locks for the entire 4GB address space), so SPU reservations often mislead the RSX that its updating FIFO bit in reality it operates on different memory entirely.

Reply:
From profiling data the issue we have is waiting for reservations to clear when we have conflict. All this code does is improve concurrency by not fighting with CELL at all if we have already have 4+ cache lines loaded. This saves time in the end and gives a minor performance uplift without breaking accuracy. The check is done on the first reservation failure which keeps the buffer in-order. On realhw the puller does not wait for the pusher to analyze and push the whole 1KB before accessing it. It only reads upto 1KB ahead when the puller is too slow to catch up (remember on realhw the puller will not be fast because commands are executing concurrently). The cache is more like a reservoir. The reason atomic FIFO works has nothing to do with the caching, but more to do with the fact that we're respecting the cache-line reservation.

Reply:
Well we don't have exact hw tests for it so we can't know when it is accurate and when it is inaccurate which is why I simply fetch the entire 1kb at all times. Better to not take risk, this option already works so well in all games, why take the chance to break it in the first place if we have a proper solution.

Reply:
SPU reservations have nothing to do with RSX FIFO in games, this is about SPU PUT command. (a memory write). SPU reservations are for read+write operation which make 0 sense for games to use for FIFO updates.

Reply:
We can improve things if it breaks games, though that makes little sense here. We aren't really reading 1KB of memory, we're just reading 128 bytes 8 times, a loop that can be broken out over time with the same effect that this PR introduces. Also, we have documentation for other NVIDIA GPUs in that generation, RSX is not unique in this behavior. Technically our whole FIFO setup is completely backwards and incorrect, but it works well enough that we keep using it as-is. The correct setup would have 2 FIFO threads: 1 to fetch cache lines, process commands and push them (including checking for invalid method and memory faults), one to pull in parallel from the cache being generated by the first one and execute. That's how the real thing is set up which is why I mention that it does not actually fill the buffer before starting to fetch. It is also possible to switch contents at any time with GPUs of this era, even with a partially drained FIFO cache according to nouveau documentation.

Reply:
It matters because when RSX is at place X, it can fetches let's say 1kb and when we have REF command update in the middle (real example) which informs the game that the RSX already prefetched the dat it can overwrite it without worries.

Reply:
Having a pipeline bubble that severe doesn't seem plausible to me. You need to remember the pusher doesn't push the whole memory block at once, it fetches a command at a time from memory and pushes a command at a time to the cache. Surely you see the design problem here with your proposal, right?